R/tt_pos_and_access.R

```diff
@@ -939,42 +1024,46 @@ setMethod("[", c("VTableTree", "numeric", "numeric"),
           function(x, i, j, ..., drop = FALSE) {
     ## have to do it this way because we can't add an argument since we don't
     ## own the generic declaration
-    keep_topleft <- list(...)[["keep_topleft"]] ## returns NULL if not presesnt
+    keep_topleft <- list(...)[["keep_topleft"]] ## returns NULL if not present
     keep_titles <- list(...)[["keep_titles"]] %||% FALSE
+    keep_footers <- list(...)[["keep_footers"]] %||% keep_titles
     reindex_refs <- list(...)[["reindex_refs"]] %||% TRUE
-    if(is.null(keep_topleft))
-        keep_topleft <- NA
-
+    
+    # browser()
     nr <- nrow(x)
     nc <- ncol(x)
     i <- .j_to_posj(i, nr)
     j <- .j_to_posj(j, nc)
 
     ##  if(!missing(i) && length(i) < nr) {
-    if(length(i) < nr) { ## already populated by .j_to_posj
-        keep_topleft <- isTRUE(keep_topleft)
+    if(length(i) < nr) {## already populated by .j_to_posj
+        if (is.null(keep_topleft)) keep_topleft <- FALSE
         x <- subset_by_rownum(x, i,
                               keep_topleft = keep_topleft,
-                              keep_titles = keep_titles)
-    } else {
-        keep_topleft <- !identical(FALSE, keep_topleft)
+                              keep_titles = keep_titles,
+                              keep_footers = keep_footers)
+    } else if (is.null(keep_topleft)) {
+        keep_topleft <- TRUE
     }
+    
+    # browser()
     ##  if(!missing(j) && length(j) < nc)
     if(length(j) < nc)
         x <- subset_cols(x, j,
                          keep_topleft = keep_topleft,
-                         keep_titles = keep_titles)
-
-    if(length(j) == 1L &&
-       length(i) == 1L &&
-       drop) {
-        rw <- collect_leaves(x, TRUE, TRUE)[[1]]
-        if(is(rw, "LabelRow"))
-            x <- NULL
-        else
-            x <- row_values(rw)[[1]]
-    }
-    if(!drop) {
+                         keep_titles = keep_titles,
+                         keep_footers = keep_footers)
+
+    # Dropping everything
+    if (drop && any(length(j) == 1L,
+                    length(i) == 1L)) {
+        rw <- collect_leaves(x, TRUE, TRUE)
```

Comment:
this looks like a behavior change, it no longer has the `[[1]]`, is that intentional?

Reply:
Yes, it is as we discussed before. In the previous comment, you wanted to have `drop` behave as it works for `data.frame`. In the following, the two behaviors are compared.
```r
tbl <- basic_table() %>%
  split_cols_by("Species") %>%
  analyze("Sepal.Length", function(x) {
    list("m" = mean(x), "sd" = sd(x))
  },
  format = "xx.x"
  ) %>%
  build_table(iris) %>%
  print()
```
produces the following:
```
     setosa   versicolor   virginica
————————————————————————————————————
m     5.0        5.9          6.6   
sd    0.4        0.5          0.6  
```

Classical `data.frame` and `rtables` produces:
```
> df <- as.data.frame(matrix_form(tbl)$string[-1,][,-1], row.names = c("m", "sd"))
> df
    V1  V2  V3
m  5.0 5.9 6.6
sd 0.4 0.5 0.6
> df[, 1, drop = FALSE]
    V1
m  5.0
sd 0.4
> tbl[, 1, drop = FALSE]
     setosa
———————————
m     5.0  
sd    0.4 

> df[, 1, drop = TRUE]
[1] "5.0" "0.4"
> tbl[, 1, drop = TRUE]
[1] 5.0060000 0.3524897
```
I modified this behavior (i.e. having `drop` working also with vectors) after our discussions. You suggested in your comment to use the same behavior that data.frame-drop uses. I added the error in the first place because I got confused by the previous printing of the table in cases where an entire or partial column or row was in input for the drop. I believe it is better to adopt the same behavior of default data.frame-drop here and allow vectors to be dropped

Reply:
I'm not sure we can do the vector for a column thing because we have to think about what happens in the case of label rows. WE'd either get a list with NULL entries in it (likely not particularly useful for the end user) or we'd get something where `NROW(tbl[,1, drop = TRUE]) != NROW(tbl)` which definitely seems like something we don't want.

Reply:
Let me test these specific cases and get back to you. I did not think about that. Intuitively I would give back only numeric values in general as drop drops everything but in cases where there are label rows I would give back the UNdropped table with a warning. Do you think it is a reasonable behavior? What about when we have mixed output numeric, characters? Again, intuitively I would give back a mixed vector or list with "" for the label rows, so we are aligned with the tabulation principles that rtables has. In the next comment, I will show you both these scenarios and some possible solutions for these specific cases

Reply:
I don't think we should do anything that doesnt' follow the `length(tbl[,i, drop = TRUE]) == NROW(tbl)` criteria. Its too likely to cause bugs or otherwise be a gotcha to be useful, in my opinion.

In practice that means very often ("almost always") giving back a list vector in the "extract a column" case, which I don't think will ultimately be very useful to people, because most non-trivial tables will have either label rows or content rows somewhere in their makeup, and content rows usually have multi-valued cells.

Reply:
you are very right. I did not think well about that. Now it does return a vector only if the one selected cell has multiple values or the row/column does not have any label rows or multiple values in it. If the last two cases are tried, a specific warning for both is produced and the dropping does not happen (i.e. it keeps the table structure). What do you think? All expected cases are tested in the new tests